x/airdrop/keeper/claim.go

```diff
@@ -0,0 +1,195 @@
+	// claimed += distributedRewards
+	userAllocation.Claimed = userAllocation.Claimed.Add(distributedRewards)
+
+	// Update the amount forfeited on the allocation record
+	// forfeited += totalAccruedRewards - distributedRewards
+	// Note: forfeited should be zero before the next operation,
+	// but we're doing += just in case there's a scenario where it's not zero in the future
+	forfeitedRewards := totalAccruedRewards.Sub(distributedRewards)
+	userAllocation.Forfeited = userAllocation.Forfeited.Add(forfeitedRewards)
+
+	// Flag the user's claim type decision
+	userAllocation.ClaimType = types.CLAIM_EARLY
+
+	// Distribute rewards from the distributor, deducting the early penalty
+	distributorAccount := sdk.MustAccAddressFromBech32(airdrop.DistributionAddress)
+	claimerAccount := sdk.MustAccAddressFromBech32(userAllocation.Address)
+
+	rewardsCoin := sdk.NewCoin(airdrop.RewardDenom, distributedRewards)
+	if err := k.bankKeeper.SendCoins(ctx, distributorAccount, claimerAccount, sdk.NewCoins(rewardsCoin)); err != nil {
+		return errorsmod.Wrapf(err, "unable to distribute rewards")
+	}
+
+	// Update the reward record for to mark the progress
+	k.SetUserAllocation(ctx, userAllocation)
+
+	return nil
+}
+
+// Admin transaction to merge allocations between a stride and non-stride address
+// If the stride address does not yet have an allocation, the host allocation will be overwritten
+// with the stride address
+// Otherwise, if the stride allocation already exists, the two allocations will be merged and set
+// in on the stride allocation
+// We can safely change the type back to DAILY because if a user claimed early, their allocations
+// will be set to 0 (they will have no remaining allocations)
+// There's no need to merge the Claimed or Forfeited amounts because the host allocations cannot
+// be claimed through a non-stride address
+func (k Keeper) LinkAddresses(ctx sdk.Context, airdropId, strideAddress, hostAddress string) error {
+	// Fetch the airdrop and user's allocations
+	_, airdropFound := k.GetAirdrop(ctx, airdropId)
+	if !airdropFound {
+		return types.ErrAirdropNotFound.Wrapf("airdrop %s", airdropId)
+	}
+	hostAllocation, hostFound := k.GetUserAllocation(ctx, airdropId, hostAddress)
+	if !hostFound {
+		return types.ErrUserAllocationNotFound.Wrapf("user %s for airdrop %s", hostAddress, airdropId)
+	}
+	strideAllocation, strideFound := k.GetUserAllocation(ctx, airdropId, strideAddress)
+
+	// If the stride user doesn't exist yet, just update the address in the host allocation
+	// to the the stride address overwrite it
+	// Also reset the claim type to DAILY
+	if !strideFound {
+		hostAllocation.Address = strideAddress
+		hostAllocation.ClaimType = types.CLAIM_DAILY
+		k.RemoveUserAllocation(ctx, airdropId, hostAddress)
+		k.SetUserAllocation(ctx, hostAllocation)
+		return nil
+	}
+
+	// Confirm the stride and host allocations are the same length
+	if len(strideAllocation.Allocations) != len(hostAllocation.Allocations) {
```

Comment:
Aren't all allocations the same length?

Reply:
it is in there when allos are initialized, this is more just a redundant check since we're about to iterate through each list